Container Dirty events documentation (#5878)
microsoft/FluidFramework: packages/loader/container-loader/README.md

## Expectations from container runtime and data store implementers

1. Respect ["readonly" state](#Readonly-states). In this state container runtime (and data stores) should not allow changes to local state, as these changes will be lost on container being closed.
2. Maintain Ops in flight until observed they are acknowledged by server. Resubmit any lost Ops on reconnection. This is done by DDSes in stock implementations of container & data store runtimes provided by Fluid Framework
3. Respect "["disconnected" and "connected"](#Connectivity-events) states and do not not submit Ops when disconnected.
4. Respect ["dispose"](#Closure) event and treat it as combination of "readonly" and "disconnected" states. I.e. it should be fully operatable (render content), but not allow edits. This is equivalent to "closed" event on container for hosts, but is broader (includes container's code version upgrades).

> 5. Notify Container about presence or lack of unsaved changed though IContainerContext.updateDirtyContainerState callback. This is required for [Dirty events](#Dirty-events) to correctly represent state of user edits to host.

Reviewer: ```suggestion
5. Notify Container about presence or lack of unsaved changed through IContainerContext.updateDirtyContainerState callback. This is required for [Dirty events](#Dirty-events) to correctly represent state of user edits to host.
```

### `storageOnly`
Storage-only mode is a readonly mode in which the container does not connect to the delta stream and is unable to submit or recieve ops. This is useful for viewing a specific version of a document.

## Dirty events

> Container runtime (IRuntime) communicates with Container about state of the container. In response, Container will raise two events - `"dirty"` and `"saved"` events. Transitions between twese two events signify presence (or lack of) user changes that were not saved to storage. In other words, if container is dirty, closing it at that moment will result in data loss from user perspective, as not all user changes made it to storage.

Reviewer: ```suggestion
The Container runtime can communicate with the container to get the container's current state. In response, the container will raise two events - `"dirty"` and `"saved"` events. Transitions between these two events signify presence (or lack of) user changes that were not saved to storage. In other words, if container is dirty, closing it at that moment will result in data loss from user perspective, because not all user changes made it to storage.
```

> This information can be used by host to build appropriate UX that allows user to be confident in the platform. For example, host may chose to show "Your changes are not saved. Are you sure you want to close document/tab/...?" kind of dialong on closing. instead (or in addition to) host may chose to show "Saving..." and "Saved" text somewhere in UX. Coupled with lack of connectvity to ordering servive (and appropriate notificaiton to the user) that may creeate enough continuos notificaiton to user not to require blocking dialog on closing.

Reviewer: ```suggestion
This information can be used by a host to build appropriate UX that allows user to be confident in the platform. For example, a host may chose to show a dialog asking the user if they want to save their changes before closing. Instead of, or in addition to this, the host may choose to show "Saving..." and "Saved" text somewhere in UX. Coupled with lack of connectivity to ordering service (and appropriate notification to the user) that may create enough continuous notification to user not to require a blocking dialog on closing.
```

> Please note that when active conneciton is in place, it's just a matter of time for changes to be flushed to storage (unless there is some source of continuous local changes being generated that prevents container from ever being fully saved). But if there is no active connection (let's say user is offline, or service experiences issues), then document may stay in dirty state for very long time.

Reviewer: ```suggestion
Note that when an active connection is in place, it's just a matter of time before changes will be flushed to storage unless there is some source of continuous local changes being generated that prevents container from ever being fully saved. But if there is no active connection, because the user is offline, for example, then a document may stay in a dirty state for very long time.
```
